cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/operations/InMemoryOperationRunner.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.operations;
+
+import com.google.inject.Inject;
+import io.cdap.cdap.api.service.operation.LongRunningOperation;
+import io.cdap.cdap.api.service.operation.LongRunningOperationRequest;
+import io.cdap.cdap.proto.id.OperationRunId;
+import io.cdap.cdap.proto.operationrun.OperationError;
+import io.cdap.cdap.proto.operationrun.OperationMeta;
+import io.cdap.cdap.proto.operationrun.OperationType;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of {@link OperationRunner} to run an operation in the same service.
+ */
+public class InMemoryOperationRunner<T> extends AbstractOperationRunner<T> {
+
+  private final OperationStatePublisher statePublisher;
+  private LongRunningOperation<T> currentOperation;
+  private final Lock lock;
+  private final ExecutorService executor;
+  private static final Logger LOG = LoggerFactory.getLogger(InMemoryOperationRunner.class);
+
+  /**
+   * Default constructor.
+   */
+  @Inject
+  public InMemoryOperationRunner(Map<OperationType, LongRunningOperation> operations,
+      OperationStatePublisher statePublisher) {
+    super(operations);
+    this.statePublisher = statePublisher;
+    this.lock = new ReentrantLock();
+    this.executor = Executors.newSingleThreadExecutor();
```

Comment:
Executor should always be created with a thread factory so that threads created underneath are named. It is important to name thread for easy debugging.

Also, where do we shutdown this executor? If we don't shutdown, it can leave behind threads that block the termination of the JVM process.

Reply:
I see the shutdown happens at the stop method. For class that create resources in constructor (like in this case), it is better to implement `AutoCloseable` instead of a custom stop method.